# Calibration Model 02: Generátor syntetických dat (Synthetic Data Generator)

### Počet kalibrovaných parametrů: 3

In [1]:
# Instalace potřebných knihoven
#%pip install pandas
#%pip install numpy
#%pip install seaborn matplotlib
#%pip install pymc
#%pip install arviz
#%pip install ipywidgets
#%pip install jupyterlab_widgets
#%pip install pytensor
#%pip install ipywidgets jupyterlab_widgets

In [2]:
# Import potřebných knihoven
import pandas as pd
import numpy as np

import pymc as pm
import arviz as az
import pytensor.tensor as pt

import seaborn as sns
import matplotlib.pyplot as plt

### Vstupní data

In [3]:
### Načtení z formátu netCDF

# Soubor je načten a přiřazen do proměnné ‚trace‘
other_path = '../../data/05_Calibration/posterior_trace_three.nc'
trace = az.from_netcdf(other_path)

In [4]:
az.summary(trace)

mean     sd  hdi_3%  hdi_97%  mcse_mean  mcse_sd  ess_bulk  \
v_load    0.651  0.111   0.460    0.854      0.002    0.001    3958.0   
v_unload  1.198  0.257   0.705    1.658      0.003    0.002    5690.0   
accel     0.550  0.159   0.345    0.854      0.002    0.003    4749.0   
sigma     3.593  0.088   3.417    3.749      0.001    0.001    8339.0   

          ess_tail  r_hat  
v_load      5840.0    1.0  
v_unload    5023.0    1.0  
accel       5884.0    1.0  
sigma       7311.0    1.0

### Definice a nastavení parametrů robotického systému

In [5]:
# SPECIFIKACE TECHNOLOGICKÉHO PROCESU ZDĚNÍ

time_refer_2_refer = 20   # s, průměrná doba pohybu z referenčního bodu k referenčnímu bodu (depaletizace, verifikace)
time_mounting = 3         # s, doba manipulaci v cílové poloze (umístění prvku)
brick_thickness = 440     # mm, tloušťka zdicího prvku (Porotherm 440 Profi)
brick_height = 250        # mm, výška zdicího prvku (Porotherm 440 Profi)
brick_width = 250         # mm, šířka zdicího prvku (Porotherm 440 Profi)

# SOUŘADNICE REFERENČNÍHO BODU (nad verifikačním stolem)

refer_x = 2_000           # mm, souřadnice X referenčního bodu
refer_y = 3_500           # mm, souřadnice Y referenčního bodu
refer_z = 1_000           # mm, souřadnice Z referenčního bodu


### Definice funkce pro výpočet celkové doby pracovního cyklu

In [ ]:
def simulate_time(dist, v_load, v_unload, accel):
    """
    Funkce pro výpočet celkové doby pracovního cyklu robotického zdění.

    Parametry:
    dist (int): dráha trajektorie od referenčního bodu k cílové poloze prvku [mm]

    Návratová hodnota:
    total_time (int): celková doba pracovního cyklu [s]
    """
    # výpočet času potřebného k dosažení maximální rychlosti
    time_to_load_speed = v_load / accel
    dist_to_load_speed = (1/2) * accel * time_to_load_speed  # uražená dráha při akceleraci

    # výpočet času potřebného k dosažení 0 rychlosti
    time_to_unload_speed = v_unload / accel
    dist_to_unload_speed = (1/2) * accel * time_to_unload_speed  # uražená dráha při deakceleraci

    # pevné technologické časy (manipulace a přesuny mezi pevnými body)
    total_time = time_refer_2_refer

    # pohyb s naloženým prvkem (převod mm → m)
    total_time += (dist - dist_to_load_speed) / 1_000 / v_load

    # pohyb bez zátěže (zpětný pohyb)
    total_time += (dist - dist_to_unload_speed) / 1_000 / v_unload

    # započtení akceleračních časů
    total_time += time_to_load_speed + time_to_unload_speed
    return total_time

### Vymezení pracovního rozsahu pro generování dat

In [7]:
# počet scénářů simulace
number_simulation = 1_000

# nastavení limitních hodnot rozsahu
dist_min = 2_000
dist_max = 10_000

# Vygeneruje 1000 náhodných hodnot z rovnoměrného rozdělení
dist_range = np.random.uniform(dist_min, dist_max, number_simulation)

### Empirická bootstrap distribuce zdržení (stochastické vlivy)

In [8]:
### Načtení hybridního modelu

# Soubor je načten a přiřazen do proměnné ‚stochastic_delay‘
other_path = '../../data/04_HybridModel/hybrid_model.csv'
stochastic_delay = pd.read_csv(other_path, header=0)

In [9]:
delay_samples = stochastic_delay["stochastic_delay"].values
delay_samples

array([0, 0, 0, ..., 0, 0, 0], shape=(20000,))

### Generátor synteticých dat

In [10]:
# =====================================================
# 1️ EXTRAKCE POSTERIOR VZORKŮ
# =====================================================

# Sloučení chain + draw do jedné dimenze
posterior = trace.posterior.stack(sample=("chain", "draw"))

# Extrakce parametrů
v_load_samples = posterior["v_load"].values
v_unload_samples = posterior["v_unload"].values
accel_samples = posterior["accel"].values
sigma_samples = posterior["sigma"].values

n_samples = len(v_load_samples)

In [11]:
# =====================================================
# 2️ MONTE CARLO SIMULACE
# =====================================================

simulated = []

rng = np.random.default_rng(122)

for d in dist_range:

    T_samples_for_dist = []

    for i in range(n_samples):

        T_det = simulate_time(
            dist=d,
            v_load=v_load_samples[i],
            v_unload=v_unload_samples[i],
            accel=accel_samples[i]
        )
        # bootstrap náhodná realizace
        delay = rng.choice(delay_samples)

        # šum (noise) s normálním rozdělením
        #delay = np.random.normal(0, sigma_samples[i], size=np.shape(T_det))

        T_sim = T_det + delay

        T_samples_for_dist.append(T_sim)

    simulated.append(T_samples_for_dist)

simulated_data = np.array(simulated)

In [12]:
simulated_data.shape

(1000, 12000)

Simulace obsahuje:
* 1 000 hodnot scénářů
* 12 000 Monte Carlo realizací pro každý scénář

### Základní vyhodnocení (pro každý scenář)

In [13]:
# =====================================================
# 3️ STATISTICKÉ VYHODNOCENÍ
# =====================================================

# Střední hodnota
T_mean = simulated_data.mean(axis=1)

# Směrodatná odchylka
T_std = simulated_data.std(axis=1)

# 95% interval spolehlivosti
T_lower = np.percentile(simulated_data, 2.5, axis=1)
T_upper = np.percentile(simulated_data, 97.5, axis=1)

# Medián
T_median = np.median(simulated_data, axis=1)

In [14]:
# =====================================================
# 4️ VÝSTUPNÍ SHRNUTÍ
# =====================================================

print("Průměr:", T_mean)
print("95% interval:", T_lower, "-", T_upper)
print("Směrodatná odchylka:", T_std)
print("Medián:", T_median)

Průměr: [41.18881472 33.78401544 34.03920117 47.93141265 51.53785407 50.3559627
 47.67483875 36.24340188 45.05262886 45.7876038  43.17789465 50.87235153
 36.64229103 48.50046874 40.54519368 40.18777522 47.10624879 51.97217714
 50.22567554 39.67439047 44.90404723 35.71348134 44.94196871 52.41411706
 34.06406258 39.27603757 41.15598114 51.78693239 44.87984327 33.30982492
 43.5892547  51.7221477  34.99341951 42.69741998 44.94836331 43.76756054
 48.97281188 46.9676684  47.02978764 50.50961235 35.02613584 42.58077345
 42.04297586 46.59779226 33.30640806 44.29876464 34.07555877 43.62981198
 34.22337075 35.80967769 38.66493416 41.97066761 50.88841332 50.85235992
 44.20757543 51.64746043 33.84330129 38.20950031 43.84957416 38.73096767
 41.90534169 33.01532121 49.13737215 40.12754924 35.499821   43.01698988
 36.37009355 42.41971094 46.34641963 46.07181319 50.41054529 44.9920067
 45.65337642 33.72599729 46.17802865 39.92686888 51.93587219 46.83264538
 37.53581346 49.18573735 36.40751078 44.25172

### Globální vyhodnocení (přes všechny scénáře)

In [15]:
global_distribution = simulated_data.flatten()

global_mean = global_distribution.mean()
global_std = global_distribution.std()

global_lower = np.percentile(global_distribution, 2.5)
global_upper = np.percentile(global_distribution, 97.5)

global_median = np.median(global_distribution)

In [16]:
print("Globální průměr:", global_mean)
print("Globální 95% interval:", global_lower, "-", global_upper)
print("Globální Směrodatná odchylka:", global_std)
print("Globální medián:", global_median)

Globální průměr: 42.79838425985259
Globální 95% interval: 28.95378425265056 - 105.77944492659331
Globální Směrodatná odchylka: 20.862916335554917
Globální medián: 39.111403196947634


In [17]:
results_df = pd.DataFrame({
    "dist": dist_range,
    "global_mean": global_mean,
    "global_CI_lower": global_lower,
    "global_CI_upper": global_upper,
    "global_std": global_std,
    "global_median": global_median,
    "T_mean": T_mean,
    "T_CI_lower": T_lower,
    "T_CI_upper": T_upper,
    "T_std": T_std,
    "T_median": T_median,

})

In [18]:
simulated_data

array([[ 36.26806134,  36.36762706,  36.59708521, ...,  36.73724633,
         37.02824368,  56.67432009],
       [ 29.35111575,  29.81018693,  29.91735364, ...,  30.32088978,
         29.22336492,  29.57314548],
       [ 29.57381412,  30.02131064,  30.13241465, ...,  30.52747115,
         29.47465123,  29.8017753 ],
       ...,
       [ 40.35129787,  40.23863902,  40.54028869, ...,  40.52497331,
         41.63564816,  40.86631137],
       [ 41.73184959,  41.54743713,  41.87349492, ...,  41.80561257,
        103.19342228,  42.28363332],
       [ 42.29304117,  42.079461  ,  42.41544064, ...,  42.32618993,
         43.82665441,  42.85977192]], shape=(1000, 12000))

### Generování datasetu syntetických dat

In [19]:
n_dist = simulated_data.shape[0]
n_mc = simulated_data.shape[1]

# zopakujeme každé dist 12000x
dist_long = np.repeat(dist_range, n_mc)

# rozbalíme časy
total_time_long = simulated_data.flatten()

# vytvoříme dataframe
synthetic_df = pd.DataFrame({
    "dist": dist_long,
    "total_time": total_time_long
})

In [20]:
synthetic_df.shape

(12000000, 2)

In [21]:
# náhodně podvzorkování
synthetic_df = synthetic_df.sample(500_000, random_state=122)

In [22]:
synthetic_df.head()

dist  total_time
8875038   8290.315041   44.716401
9799933   3904.932180   33.406397
3368641   7249.864288   41.944719
10520965  7470.111744   42.668021
4406391   4087.879114   33.873320

### Export datové sady do formátu CSV

In [23]:
synthetic_df.to_csv("../../data/05_Calibration/synthetic_dataset.csv", index=False)
results_df.to_csv("../../data/05_Calibration/results_synthetic_dataset.csv", index=False)

### Autor / Organizace / Datum

Vjačeslav Usmanov, ČVUT v Praze, Fakulta stavební

###### Přehled změn


|  Datum (YYYY-MM-DD) |  Verze | Autor změny  |  Popis změny |
|---|---|---|---|
| 2026-01-31 | 1.1 | Vjačeslav Usmanov| added CM_03_SyntGenerator.ipynb |
| 2026-02-18 | 1.2 | Vjačeslav Usmanov| changed CM_03_SyntGenerator.ipynb |